File: easy/9-PalindromeNumber.js
// This version turns it into a string first
// /**
//  * @param {number} x
//  * @return {boolean}
//  */
// var isPalindrome = function (x) {
//   if (x < 0) {
//     return false;
//   }
//   let str = x.toString();
//   for (let i = 0; i < str.length / 2; i++) {
//     if (str[i] != str[str.length - 1 - i]) {
//       return false;
//     }
//   }
//   return true;
// };

// This version does not turn it into a string first
/**
 * @param {number} y
 * @returns {boolean}
 */
var getReversedInt = function (y) {
  let digit,
    reversedY = 0;
  do {
    digit = y % 10;
    reversedY = reversedY * 10 + digit;
    y = (y - digit) / 10;
  } while (y > 0);
  return reversedY;
};

/**
 * @param {number} x
 * @return {boolean}
 */
var isPalindrome = function (x) {
  if (x < 0) {
    return false;
  }
  return x == getReversedInt(x);
};

console.log(isPalindrome(121), true);
console.log(isPalindrome(-121), false);
console.log(isPalindrome(10), false);

console.log(isPalindrome(1278721), true);
console.log(isPalindrome(1278821), false);
console.log(isPalindrome(1), true);

File: easy/9-PalindromeNumber.c
#include <stdbool.h>
#include <stdio.h>

// This version turns it into a string first
// bool isPalindrome(int x) {
//   if (x < 0) {
//     return false;
//   }

//   char str[12];
//   int strLen = sprintf(str, "%d", x);

//   for (int i = 0; i < strLen / 2; i++) {
//     if (str[i] != str[strLen - 1 - i]) {
//       return false;
//     }
//   }

//   return true;
// }

// This version does not turn it into a string first
int getReversedInt(int y) {
  int digit = 0;
  long reversedY = 0;
  do {
    digit = y % 10;
    reversedY = reversedY * 10 + digit;
    y = (y - digit) / 10;
  } while (y > 0);
  printf("%ld", reversedY);
  return reversedY;
}

bool isPalindrome(int x) {
  if (x < 0) {
    return false;
  }

  return x == getReversedInt(x);
}

int main(int argc, char** argv) {
  printf("%s %s\n", isPalindrome(121) ? "true" : "false", "true");
  printf("%s %s\n", isPalindrome(-121) ? "true" : "false", "false");
  printf("%s %s\n", isPalindrome(10) ? "true" : "false", "false");

  printf("%s %s\n", isPalindrome(1278721) ? "true" : "false", "true");
  printf("%s %s\n", isPalindrome(1278821) ? "true" : "false", "false");
  printf("%s %s\n", isPalindrome(1234567899) ? "true" : "false", "false");

  return 0;
}
